@alloy => [FilterArtworks] Spike onto alternate schema design for V2

mzikherman opened the pull request:
This adds a new type, `filterArtworksConnection`.

It is fundamentally a connection of artworks (ie- the *nodes* are plain artworks), but the connection-level fields are enhanced w/ all of the usual suspects like `aggregations`, etc.

This definitely feels nicer! Here's a sample query:

```
{
  filterArtworksConnection(keyword: "warhol", first: 30, after: "YXJyYXljb25uZWN0aW9uOjI5", aggregations: [MEDIUM,TOTAL]) {
    aggregations {
      slice
    }
    edges {
      node {
        slug
      }
    }
  }
}
```

mzikherman commented on src/schema/v2/filterArtworksConnection.ts:
```diff
@@ -0,0 +1,134 @@
+
+// The connection fields here match the original implementation's,
+// except without `artworks_connection`.
+// TODO: Support proper thunk.
```
In order to DRY it up, I'm trying to re-use the fields already defined for filter artworks, but removing the `artworks_connection` one. However, that field definition is a thunk (avoiding some circular/diamond dependency stuff), and I can't quite figure out how to share it in this way (since here I'm calling it directly).

Kind of #minor since I can just copy/paste the fields! More just nerd-sniping.

alloy replied:
Perhaps?

```ts
const thunk = () => {
  const { artworks_connection, ...connectionFields } = FilterArtworksFields()
  return connectionFields
}
```

But copy-pasting might be preferable, or move the definitions here and refer to them from the deprecated version instead? 🤔

mzikherman commented on src/schema/v2/filterArtworksConnection.ts:
```diff
@@ -0,0 +1,134 @@
+const filterArtworksConnectionType = connectionDefinitions({
+  name: "FilterArtworks",
+  nodeType: Artwork.type,
+  // interfaces: [NodeInterface],
```
This seems like possibly an update in the lib (it can accept all of the object configs, besides the specific connection ones, and pass those through).

Or, can maybe just be used via `filterArtworksConnectionType.interfaces = [NodeInterface]`. I didn't play with this really.

alloy replied:
Remind me, this is so we can refer to the connection and its filter state by a global ID?

You can’t mutate a graphql-js type, so the latter wouldn’t work.

mzikherman replied:
Yup! I'm wondering if we actually need this though, or if a node query would resolve properly anyway (will check it out).

If it's a blocker then I can patch-package the lib to allow more options to be passed in to the GraphQL object type created.

alloy commented:
Agreed, feels nicer! 👌

alloy commented:
In order to move this forward, could you perhaps try to make use of it in Emission to see if everything works well in practice? I think this is the only place where we use filtered artworks https://github.com/artsy/emission/blob/a5cb8c826541e123b484569902eaad400a838c3e/src/lib/Components/FilteredInfiniteScrollGrid/FilteredInfiniteScrollGrid.tsx

mzikherman commented:
Whoops @alloy sorry I missed your messages!

Will clean up this PR and get it mergeable today (can check out Emission too).

mzikherman commented:
Now with node id support:

```
{
  node(id:"ZmlsdGVyQXJ0d29ya3NDb25uZWN0aW9uOnsicGFnZSI6MSwic2l6ZSI6MTAsIm9mZnNldCI6MCwiYWdncmVnYXRpb25zIjpbInRvdGFsIiwibWVkaXVtIl0sImtleXdvcmQiOiJhbmR5In0=") {
    ... on FilterArtworksConnection {
      id
      edges {
          node {
          slug
        }
      }
    }
  }
}
```

mzikherman commented:
Ok, I think this is ready for review/mergeable. The normal use cases are working, as well as a node interface implementation for the connection (using patch-package).

Was adding some tests, but realized we're not running tests for V2: https://github.com/artsy/metaphysics/blob/4b51fbf3a503f4afd415b5450d114b1ed7575fcb/package.json#L159

I can comment that line out and commit some working tests in a `filterArtworksConnection.test.js` if you think that's helpful (and then comment that line back in).

alloy commented:
> I can comment that line out and commit some working tests in a `filterArtworksConnection.test.js `if you think that's helpful (and then comment that line back in).

Sure, yeah, if that works then please do 👍 @zephraph is fixing the test suite at the moment, so we’ll soon be running the full suite again.

alloy commented on src/schema/v2/filterArtworksConnection.ts:
```diff
@@ -0,0 +1,152 @@
+    const { include_artworks_by_followed_artists, aggregations } = options
+    const requestedPersonalizedAggregation = aggregations.includes(
+      "followed_artists"
+    )
```
I think you’re missing the camelCase args to snake_case for Gravity munging here. See e.g. https://github.com/artsy/metaphysics/blob/4b51fbf3a503f4afd415b5450d114b1ed7575fcb/src/schema/v2/gene.ts#L90-L137

mzikherman replied:
Ah, indeed.

mzikherman commented:
Ok @alloy , added some passing specs (which currently only run locally, not on CI), and addressed the incoming args camel -> snake (for the proper API call) issue.

alloy commented:
Perfect thanks, going to give this a go in Emission now!

alloy commented:
@mzikherman So I’ve gone ahead and dropped all legacy filter artwork usage in favour of this new connection, which we’ll be QAing in the app today.
